Hello Michael!

My name is Elena. I'm happy to review your project today. 

When I will see mistake at the first time, I will just point it out. I let you find it and fix it by yourself. I'm trying to prepare you to work as an Data Analyst. SO, at a real job, your team lead will do the same. But if you can't handle this task yet, I will give you a more accurate hint at the next iteration.

Below you will find my comments - please do not move, modify or delete them.

You can find my comments in green, yellow or red boxes like this:

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Success. Everything is done succesfully.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Remarks. Some recommendations.
</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>

You can answer me by using this: 

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Final reviewer's comment</b> <a class="tocSkip"></a>

You've done a great job! 
    
It was my pleasure to check your work:
1. There is a structure;
2. You add subtotals;
3. Well described the Final conclusions at the end of the project;
4. Correctly removed duplicates and made data categorization;
5. Correctly replaced the missing values in the "total_income" column;
6. He argued his conclusions in the Checking the Hypotheses block.    
</div>

<div class="alert alert-block alert-warning">
<b>Final reviewer's comment</b> <a class="tocSkip"></a>

Please pay attention to a few nuances:
1. It is desirable to add a block describing the Brief Work Plan to the project description;
2. It is better to load all libraries at the very beginning of the project in one cell, and read the project in the next cell;
3. For a quick dive into Python, use the documentation. It will help you optimize your code and take you to the next level.    
</div>

<div class="alert alert-block alert-danger">
<b>Final reviewer's comment</b> <a class="tocSkip"></a>

~~But there are a one remark that need to be completed so that the work can be accepted:~~
~~ - Please fix the method of replacing incorrect values in the "days_employed" column.~~
    
If you have any questions - write, I will definitely help to solve them.
    
I know you will succeed!
</div>

<div class="alert alert-block alert-success">
<b>Final reviewer's comment  2</b> <a class="tocSkip"></a>
Thank you for making changes to the project. You've done excellent research.
</div>

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.

[In this notebook you're provided with hints and brief instructions and thinking prompts. Don't ignore them as they are designed to equip you with the structure for the project and will help you analyze what you're doing on a deeper level. Before submitting your project, make sure you remove all hints and descriptions provided to you. Instead, make this report look as if you're sending it to your teammates to demonstrate your findings - they shouldn't know you had some external help from us! To help you out, we've placed the hints you should remove in square brackets.]

[Before you dive into analyzing your data, explain the purposes of the project and hypotheses you're going to test.]

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
You did the right thing by adding a description of the projects and purpose.
This is a good indicator of the quality work of the Data Analyst.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
I would also suggest adding a Short Plan of Work.
A short plan of work - here, highlight the points and, if necessary, sub-points, what actions you are going to perform. With it, it will be more convenient for you to complete the project yourself and know which stages you have already completed, and which ones you need to add.
This will allow you to reach a new level and show future employers that you know what to do, you know how you will do it, and understand why. I am sure that you will succeed!
</div>

## Open the data file and have a look at the general information. 

[Start with importing the libraries and loading the data. You may realise that you need additional libraries as you go, which is totally fine - just make sure to update this section when you do.]

In [1]:
# Loading all the libraries
import pandas as pd
import numpy as np
import nltk

from nltk.stem import WordNetLemmatizer
from nltk.stem import SnowballStemmer    
from pymystem3 import Mystem
from collections import Counter

In [2]:
# Load the data
try: 
    credit_scoring = pd.read_csv('credit_scoring_eng.csv')
except:
    credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
#print(credit_scoring.head())   

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Well done! Using try - except is good practice.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
It's good practice to split cells: 1st cell with library loading and 2nd cell with dataset reading. Please pay attention to this!
</div>

## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan

[Now let's explore our data. You'll want to see how many columns and rows it has, look at a few rows to check for potential issues with the data.]

In [3]:
# Let's see how many rows and columns our dataset has

credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
# let's print the first N rows

print(credit_scoring.head(10))  

   children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1

[Describe what you see and notice in your printed data sample. Are there any issues that may need further investigation and changes?]
days employed has negative values which is ilogical, education has upper case and lower case values that are duplicates, purpose is a bit wonky as some of the values are essentialy the same thing with diferent words and may need steeming.  

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Well done! Good intermediate conclusions.
</div>

In [5]:
# Get info on data
credit_scoring.info()
credit_scoring.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

[Are there missing values across all columns or just a few? Briefly describe what you see in 1-2 sentences.]
'days_employed' has lots of mising values, as much as 'total_income'. No missing values in the other columns.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Good! Missing values are correctly defined.
</div>

In [6]:
# Let's look at the filtered table with missing values in the the first column with missing data
credit_scoring[credit_scoring.days_employed.isnull()]
credit_scoring_nan = credit_scoring[credit_scoring.days_employed.isnull()].sort_values(by = 'income_type')
#credit_scoring_nan.groupby('income_type')['income_type'].count() / credit_scoring.groupby('income_type')['income_type'].count()
#credit_scoring[credit_scoring.days_employed.isnull()]['family_status'].value_counts()
#credit_scoring[credit_scoring.days_employed.isnull()]['purpose'].value_counts()


[Do missing values seem symmetric? Can we be sure in this assumption? Explain your thoughts briefly in this section. You may probably want to conduct further investigations, and count the missing values in all the rows with missing values to confirm the the missing samples are of the same size.]

They do the missing values are simetric where the data on 'days_employed' is missing the data on 'total_income' is also missing. The missing data does not show any conection to 'income_type' (unemployed and students have non null values), 'purpose' or 'family_status' as far as the data grouping shows. 

In [7]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.
#credit_scoring_nan.info()
#credit_scoring.groupby('income_type').count()
credit_scoring_nan.loc[:, 'purpose'].count()/credit_scoring.loc[:, 'purpose'].count()
#10% of the data is missing
credit_scoring[(credit_scoring.days_employed.isnull())&(credit_scoring.total_income.isnull())]
#Data is symetrical

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

**Intermediate conclusion**

[Does the number of rows in the filtered table match the number of missing values? What conclusion can we make from this?]
That were not dealing with randomly missing data but with a patern.

[Calculate the percentage of the missing values compared to the whole dataset. Is it a considerably large piece of data? If so, you may want to fill the missing values. To do that, firstly we should consider whether the missing data could be due to the specific client characteristic, such as employment type or something else. You will need to decide which characteristic *you* think might be the reason. Secondly, we should check whether there's any dependence missing values have on the value of other indicators with the columns with identified specific client characteristic.]
About 10% of the data is missing, but the missing data does not show any conection to 'income_type' (unemployed and students for example have non null values), 'purpose' or 'family_status' as far as the data grouping shows. The only thing symetric to the missing data is the 'total_income' column.

[Explain your next steps and how they correlate with the conclusions you made so far.]
I suppose that by finding median income and mean days of employment the missing values could be restored but seeing the the 'days_employed' has both negative (which is ilogical) and aberant ( 340266 days of employment are about 931 years, which is ilogical), I'm tempted to dump those rows.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
You correctly identified the problems in the data. Keep it up!
</div>

In [8]:
# Let's investigate clients who do not have data on identified characteristic and the column with the missing values
credit_scoring_nan


children  days_employed  dob_years            education  education_id  \
10314         0            NaN         23  secondary education             1   
13710         0            NaN         26  secondary education             1   
4979          0            NaN         55  secondary education             1   
5009          0            NaN         45  secondary education             1   
5014          0            NaN          0  secondary education             1   
...         ...            ...        ...                  ...           ...   
15021         0            NaN         60    bachelor's degree             0   
13835         0            NaN         64  secondary education             1   
3907          0            NaN         56  secondary education             1   
3809          0            NaN         66  secondary education             1   
12            0            NaN         65  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
10314          unmarried                 4      F    business     0   
13710            married                 0      F    business     0   
4979             married                 0      F    business     0   
5009            divorced                 3      F    business     0   
5014             married                 0      F    business     0   
...                  ...               ...    ...         ...   ...   
15021            married                 0      F     retiree     1   
13835    widow / widower                 2      F     retiree     0   
3907             married                 0      F     retiree     0   
3809             married                 0      F     retiree     1   
12     civil partnership                 1      M     retiree     0   

       total_income                              purpose  
10314           NaN                      buy real estate  
13710           NaN               building a real estate  
4979            NaN                 university education  
5009            NaN                 getting an education  
5014            NaN                      buy real estate  
...             ...                                  ...  
15021           NaN                                 cars  
13835           NaN                 getting an education  
3907            NaN  purchase of the house for my family  
3809            NaN     transactions with my real estate  
12              NaN                    to have a wedding  

[2174 rows x 12 columns]

In [9]:
# Checking distribution
credit_scoring_nan['income_type'].value_counts()
credit_scoring_nan.describe(include='all').T

count unique                  top  freq       mean       std  \
children          2174    NaN                  NaN   NaN   0.552438   1.46936   
days_employed        0    NaN                  NaN   NaN        NaN       NaN   
dob_years         2174    NaN                  NaN   NaN     43.632   12.5315   
education         2174     12  secondary education  1408        NaN       NaN   
education_id      2174    NaN                  NaN   NaN   0.800828  0.530157   
family_status     2174      5              married  1237        NaN       NaN   
family_status_id  2174    NaN                  NaN   NaN   0.975161   1.41822   
gender            2174      2                    F  1484        NaN       NaN   
income_type       2174      5             employee  1105        NaN       NaN   
debt              2174    NaN                  NaN   NaN  0.0781969  0.268543   
total_income         0    NaN                  NaN   NaN        NaN       NaN   
purpose           2174     38     having a wedding    92        NaN       NaN   

                  min   25%  50%  75%  max  
children           -1     0    0    1   20  
days_employed     NaN   NaN  NaN  NaN  NaN  
dob_years           0    34   43   54   73  
education         NaN   NaN  NaN  NaN  NaN  
education_id        0  0.25    1    1    3  
family_status     NaN   NaN  NaN  NaN  NaN  
family_status_id    0     0    0    1    4  
gender            NaN   NaN  NaN  NaN  NaN  
income_type       NaN   NaN  NaN  NaN  NaN  
debt                0     0    0    0    1  
total_income      NaN   NaN  NaN  NaN  NaN  
purpose           NaN   NaN  NaN  NaN  NaN

[Describe your findings here.]

**Possible reasons for missing values in data**

[Propose your ideas on why you think the values might be missing. Do you think they are missing randomly or there are any patterns?]
Were not dealing with randomly missing data but with a patern. The simtric nature of the data shows that clearly.
[Let's start checking whether the missing values are random.]

In [10]:
# Checking the distribution in the whole dataset

credit_scoring.describe(include='all').T

count unique                  top   freq       mean  \
children          21525    NaN                  NaN    NaN   0.538908   
days_employed     19351    NaN                  NaN    NaN    63046.5   
dob_years         21525    NaN                  NaN    NaN    43.2934   
education         21525     15  secondary education  13750        NaN   
education_id      21525    NaN                  NaN    NaN   0.817236   
family_status     21525      5              married  12380        NaN   
family_status_id  21525    NaN                  NaN    NaN   0.972544   
gender            21525      3                    F  14236        NaN   
income_type       21525      8             employee  11119        NaN   
debt              21525    NaN                  NaN    NaN  0.0808827   
total_income      19351    NaN                  NaN    NaN    26787.6   
purpose           21525     38     wedding ceremony    797        NaN   

                       std      min      25%      50%      75%     max  
children           1.38159       -1        0        0        1      20  
days_employed       140827 -18388.9 -2747.42 -1203.37 -291.096  401755  
dob_years          12.5746        0       33       42       53      75  
education              NaN      NaN      NaN      NaN      NaN     NaN  
education_id      0.548138        0        1        1        1       4  
family_status          NaN      NaN      NaN      NaN      NaN     NaN  
family_status_id   1.42032        0        0        0        1       4  
gender                 NaN      NaN      NaN      NaN      NaN     NaN  
income_type            NaN      NaN      NaN      NaN      NaN     NaN  
debt              0.272661        0        0        0        0       1  
total_income       16475.5  3306.76  16488.5  23202.9  32549.6  362497  
purpose                NaN      NaN      NaN      NaN      NaN     NaN

**Intermediate conclusion**

[Is the distribution in the original dataset similar to the distribution of the filtered table? What does that mean for us?]
That the missing data could be dumped as it does not influance the distrebution, and it is within the limits of the data we're able to dump (10%). It probably came from people who didn't want to divulge​ the days_employed and total_income data so we're lacking it.

[If you think we can't make any conclusions yet, let's investigate our dataset further. Let's think about other reasons that could lead to data missing and check if we can find any patterns that may lead us to thinking that the missing values are not random. Because this is your work, this is section is optional.]

In [11]:
# Check for other reasons and patterns that could lead to missing values



**Intermediate conclusion**

[Can we finally confirm that missing values are accidental? Check for anything else that you think might be important here.]

In [12]:
# Checking for other patterns - explain which

**Conclusions**

[Did you find any patterns? How did you come to this conclusion?]
The simtric nature of the data shows that there's a pattern.

[Explain how you will address the missing values. Consider the categories in which values are missing.]
First I'm going to look at the mean and median values of the other values of the column and try to find what atribute or parameter, is the most influential in the distribution of mean/median values of the column to which the missing values belong. Then I'm going to input values based on the atribute that was most influential. After that I'm going to check that the formaly missing values are no longer missing.   

[Briefly plan your next steps for transforming data. You will probably need to address different types of issues: duplicates, different registers, incorrect artifacts, and missing values.]
Asses percentage of duplicates, if there are few (low pracentage) then they can be droped, but if there are many (high pracentage) they need to be raplaced.
I'll use .lower to make the regitries uniform and prevent the duplication of values.
Incorrect artifacts need to be trated based on the nature of the problem with them, different problems lead to different solutions.
As for missing values, they can be restored via mean/median based on the most infuential atribute, the one we will find based on the existing values most dominant in it's influance on the distribution of mean/median values.
I think the sphere of employment (buissness, retail, service, manufacture, etc.) will prove most dominant in the determination of the values that will be used, but in this step this is nothing more then a hypotesis. 

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

~~Please pay attention to the fact that 10% of data is a lot and you can’t just delete it. One of the important qualities of a good Data Analyst is the ability to come up with a logical and maximally correct way to fill in the missing values. Do you think the average income (or median) for a worker in one area would be considered the correct way to fill in the missing values?
Please add to this section of the project~~
</div>



<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
Thanks, I changed the conclusion, and added the section that was reqired. I hope its better now. 
</div>


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Well done! Good
</div>

## Data transformation

[Let's go through each column to see what issues we may have in them.]
days employed has negative values which is ilogical, education has upper case and lower case values that are duplicates, purpose is a bit wonky as some of the values are essentialy the same thing with diferent words and may need steeming.

[Begin with removing duplicates and fixing educational information if required.]

In [13]:
# Let's see all values in education column to check if and what spellings will need to be fixed
credit_scoring.education.unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

In [14]:
# Fix the registers if required
credit_scoring['education'] = credit_scoring.education.str.lower()

In [15]:
# Checking all the values in the column to make sure we fixed them
credit_scoring.education.unique()

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

[Check the data the `children` column]

In [16]:
# Let's see the distribution of values in the `children` column
credit_scoring.children.value_counts()
#credit_scoring[credit_scoring['children'] == -1]['children'].count()/credit_scoring.children.count()
credit_scoring[credit_scoring['children'] == 20]['children'].count()/credit_scoring.children.count()

0.0035307781649245064

[Are there any strange things in the column? If yes, how high is the percentage of problematic data? How could they have occurred? Make a decision on what you will do with this data and explain you reasoning.]
Well that -1 doesnt look logical, I meen how can you have -1 children. That 20 value also lokes fishy (I meen 20 children are a bit much), but I suppose it's possible (Mormons in Utah and other religious groups practicing polygamy, have lots of childen I think) but highly unlikely. The couse in both cases is human error, I think.
I suppose losing 0.5% (0.02 of -1, and 0.03 of 20) of the data is not a problem, besides there is no way to know what is the correct data.

In [17]:
# [fix the data based on your decision]
credit_scoring = credit_scoring[credit_scoring['children'] != -1]
credit_scoring = credit_scoring[credit_scoring['children'] != 20]

In [18]:
# Checking the `children` column again to make sure it's all fixed

credit_scoring.children.value_counts()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Well done! The "children" column is processed correctly.
</div>

[Check the data in the `days_employed` column. Firstly think about what kind of issues could there be and what you may want to check and how you will do it.]

In [19]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage
#credit_scoring.days_employed.value_counts()
credit_scoring[credit_scoring['days_employed'] < 0]['days_employed'].count() #15809 values
credit_scoring[credit_scoring['days_employed'] < 0]['days_employed'].count() / credit_scoring.children.count()
# 73.86% of values are negative which is illogical
credit_scoring[credit_scoring.days_employed.isnull()]['children'].count() #2162 values
credit_scoring[credit_scoring.days_employed.isnull()]['children'].count() / credit_scoring.children.count()
# 10.1% of values are null (as prevous parts show they are of no consiqence(it does not influance the distrebution), they can be ignored or droped)
# 70 years are about 25,568 days, which is un-reasnoble maximum (who can work for 70 years) for employment days
credit_scoring[credit_scoring['days_employed'] > 25568]['days_employed'].count() #3431 values
credit_scoring[credit_scoring['days_employed'] > 25568]['days_employed'].count() / credit_scoring.children.count()
#16.03% of values are too high
credit_scoring.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  secondary education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  secondary education             1   
8          2   -6929.865299         35    bachelor's degree             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate

[If the amount of problematic data is high, it could've been due to some technical issues. We may probably want to propose the most obvious reason why it could've happened and what the correct data might've been, as we can't drop these problematic rows.]

In [20]:
# Address the problematic values, if they exist
#credit_scoring_abs = abs(credit_scoring[credit_scoring['days_employed'] < 0]['days_employed'])
credit_scoring['days_employed'] = abs(credit_scoring['days_employed'])
#I assume that someone wrote in hours instead of days, to explain abnormaly high values (but that's only an assumption)

#def hours_to_days(days):
#    if days > 25568:
#        return days/24
#    return days

#credit_scoring_days = credit_scoring['days_employed'].apply(
#    hours_to_days
#) 
#credit_scoring['days_employed'] = credit_scoring_days

credit_scoring['days_employed']=credit_scoring['days_employed']/24
credit_scoring['days_employed'].head(20)

0       351.569709
1       167.700156
2       234.309275
3       171.864467
4     14177.753002
5        38.591076
6       119.966752
7         6.365815
8       288.744387
9        91.198185
10      173.811819
11       33.029245
12             NaN
13       76.943414
14       76.873174
15       40.515184
16       71.663926
17       98.749988
18    16678.380705
19      418.284106
Name: days_employed, dtype: float64

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

~~The assumption that the data is in hours rather than days is a great idea. But why do you think that only part of the data is in hours? Maybe it's a feature of the data? Please correct this section of the project.~~
</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
Thanks, I will corect it, but if this is a feature of the data then some values (like the ones in rows 7, 9, 11 and 15), become too low (under a month of emploment), so I think only part of the data has this problem. That leaves the question what is the norm for these values. Also then the abnormal values remain very different from the rest of the data in the column.
</div>


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
I'm glad you brought this to your attention! Well done!
</div>

In [21]:
# Check the result - make sure it's fixed
credit_scoring['days_employed'] 

0          351.569709
1          167.700156
2          234.309275
3          171.864467
4        14177.753002
             ...     
21520      188.721528
21521    14330.725172
21522       88.056120
21523      129.686738
21524       82.687816
Name: days_employed, Length: 21402, dtype: float64

[Let's now look at the client's age and whether there are any issues there. Again, think about what can data can be strange in this column, i.e. what cannot be someone's age.]

In [22]:
# Check the `dob_years` for suspicious values and count the percentage

credit_scoring.dob_years.value_counts()

35    614
40    603
41    603
34    597
38    595
42    592
33    577
39    572
31    556
36    553
29    543
44    543
30    536
48    536
37    531
43    510
50    509
32    506
49    505
28    501
45    494
27    490
52    483
56    482
47    480
54    476
46    469
58    461
57    457
53    457
51    446
55    441
59    441
26    406
60    376
25    356
61    353
62    351
63    268
64    263
24    263
23    252
65    194
66    183
22    183
67    167
21    110
0     100
68     99
69     83
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

[Decide what you'll do with the problematic values and explain why.]

In [23]:
# Address the issues in the `dob_years` column, if they exist
# drop the 0 cause todlers can't work or take loans 
credit_scoring = credit_scoring[credit_scoring['dob_years'] != 0]
credit_scoring.dob_years.value_counts()

35    614
40    603
41    603
34    597
38    595
42    592
33    577
39    572
31    556
36    553
29    543
44    543
30    536
48    536
37    531
43    510
50    509
32    506
49    505
28    501
45    494
27    490
52    483
56    482
47    480
54    476
46    469
58    461
53    457
57    457
51    446
55    441
59    441
26    406
60    376
25    356
61    353
62    351
63    268
64    263
24    263
23    252
65    194
66    183
22    183
67    167
21    110
68     99
69     83
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [24]:
# Check the result - make sure it's fixed
credit_scoring.dob_years.value_counts()

35    614
40    603
41    603
34    597
38    595
42    592
33    577
39    572
31    556
36    553
29    543
44    543
30    536
48    536
37    531
43    510
50    509
32    506
49    505
28    501
45    494
27    490
52    483
56    482
47    480
54    476
46    469
58    461
53    457
57    457
51    446
55    441
59    441
26    406
60    376
25    356
61    353
62    351
63    268
64    263
24    263
23    252
65    194
66    183
22    183
67    167
21    110
68     99
69     83
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

[Now let's check the `family_status` column. See what kind of values there are and what problems you may need to address.]

In [25]:
# Let's see the values for the column

credit_scoring.family_status.value_counts()

married              12254
civil partnership     4139
unmarried             2783
divorced              1179
widow / widower        947
Name: family_status, dtype: int64

In [26]:
# Address the problematic values in `family_status`, if they exist

#no such values as far as I can see. 'civil partnership' and 'married', are different 'de jure', so I assume different 'de facto'.

In [27]:
# Check the result - make sure it's fixed


[Now let's check the `gender` column. See what kind of values there are and what problems you may need to address]

In [28]:
# Let's see the values in the column
credit_scoring.gender.value_counts()

F      14083
M       7218
XNA        1
Name: gender, dtype: int64

In [29]:
# Address the problematic values, if they exist
#not sure, as there are non-binary in some gov. documents as a choice (but I assume in this case this is a mistake)
credit_scoring = credit_scoring[credit_scoring['gender'] != 'XNA']
credit_scoring.gender.value_counts()

F    14083
M     7218
Name: gender, dtype: int64

In [30]:
# Check the result - make sure it's fixed
credit_scoring.gender.value_counts()


F    14083
M     7218
Name: gender, dtype: int64

[Now let's check the `income_type` column. See what kind of values there are and what problems you may need to address]

In [31]:
# Let's see the values in the column
credit_scoring.income_type.value_counts()

employee                       10996
business                        5033
retiree                         3819
civil servant                   1447
entrepreneur                       2
unemployed                         2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

In [32]:
# Address the problematic values, if they exist
# entrepreneur is the same as business
#credit_scoring[credit_scoring['income_type']=='entrepreneur']['income_type'] 
credit_scoring.loc[5936, 'income_type'] = 'business'
credit_scoring.loc[18697, 'income_type'] = 'business'
#credit_scoring.income_type.value_counts()

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
See how the query method works. He will help you in your future work. Documentation link - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html
</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    Thanks.
</div>


In [33]:
# Check the result - make sure it's fixed

credit_scoring.income_type.value_counts()

employee                       10996
business                        5035
retiree                         3819
civil servant                   1447
unemployed                         2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

[Now let's see if we have any duplicates in our data. If we do, you'll need to decide what you will do with them and explain why.]

In [34]:
# Checking duplicates
credit_scoring.duplicated().sum()


71

In [35]:
# Address the duplicates, if they exist
#54 duplicates out of 21525 entries (0.25%) are insignificant thus can be droped
credit_scoring = credit_scoring.drop_duplicates().reset_index(drop = True)

In [36]:
# Last check whether we have any duplicates
credit_scoring.duplicated().sum()

0

In [37]:
# Check the size of the dataset that you now have after your first manipulations with it
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 12 columns):
children            21230 non-null int64
days_employed       19149 non-null float64
dob_years           21230 non-null int64
education           21230 non-null object
education_id        21230 non-null int64
family_status       21230 non-null object
family_status_id    21230 non-null int64
gender              21230 non-null object
income_type         21230 non-null object
debt                21230 non-null int64
total_income        19149 non-null float64
purpose             21230 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 1.9+ MB


[Describe your new dataset: briefly say what's changed and what's the percentage of the changes, if there were any.]
Droped 54 duplicated entries, addresed some of the problematic data (though not all of it the problem with 'total_income' and 'days_employed', is still there).

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Well done! Duplicates removed correctly. Columns "children", "education", "income_type", "gender" and "dob_years" are adjusted correctly. Keep it up!
</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

~~Please take note of the suggestion to take a different approach to correcting the data in the "days_employed" column and correct this section of the project.~~
</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
Thanks, I have taken a different approach there. 
I hope I've understood what was needed correctly, and that now it is better.
</div>


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Well done!
</div>

# Working with missing values

[To speed up working with some data, you may want to work with dictionaries for some values, where IDs are provided. Explain why and which dictionaries you will work with.]

In [38]:
# Find the dictionaries

### Restoring missing values in `total_income`

[Briefly state which column(s) have values missing that you need to address. Explain how you will fix them.]
'total_income' and 'days_employed' have missing values.
I suppose that by finding median income and mean days of employment the missing values could be restored.

[Start with addressing total income missing values. Create and age category for clients. Create a new column with the age category. This strategy can help with calculating values for the total income.]
categories:
'teen' < 20
20 <='young adault'< 35
35 <='adult'< 50
50 <='late middle aged'< 65
'senior citizen' > 65

In [39]:
# Let's write a function that calculates the age category
def age_group(age):
    if age < 20:
        return 'teen'
    if age < 35:
        return 'young adault'
    if age < 50:
        return 'adult'
    if age < 65:
        return 'late middle aged'
    return 'senior citizen'
    

In [40]:
# Test if the function works
print(age_group(18))
print(age_group(20))
print(age_group(32))
print(age_group(45))
print(age_group(56))
print(age_group(67))

teen
young adault
young adault
adult
late middle aged
senior citizen


In [41]:
# Creating new column based on function
credit_scoring['age_group'] = credit_scoring['dob_years'].apply(
    age_group
)
#credit_scoring['age_group'].value_counts()

In [42]:
# Checking how values in the new column
credit_scoring['age_group'].value_counts()


adult               8178
late middle aged    6229
young adault        5916
senior citizen       893
teen                  14
Name: age_group, dtype: int64

[Think about the factors on which income usually depends. Eventually, you will want to find out whether you should use mean or median values for replacing missing values. To make this decision you will probably want to look at the distribution of the factors you identified as impacting one's income.]

[Create a table that only has data without missing values. This data will be used to restore the missing values.]

In [43]:
# Create a table without missing values and print a few of its rows to make sure it looks fine
credit_scoring_full = credit_scoring.dropna()
credit_scoring_full.head(15)

children  days_employed  dob_years            education  education_id  \
0          1     351.569709         42    bachelor's degree             0   
1          1     167.700156         36  secondary education             1   
2          0     234.309275         33  secondary education             1   
3          3     171.864467         32  secondary education             1   
4          0   14177.753002         53  secondary education             1   
5          0      38.591076         27    bachelor's degree             0   
6          0     119.966752         43    bachelor's degree             0   
7          0       6.365815         50  secondary education             1   
8          2     288.744387         35    bachelor's degree             0   
9          0      91.198185         41  secondary education             1   
10         2     173.811819         36    bachelor's degree             0   
11         0      33.029245         40  secondary education             1   
13         0      76.943414         54         some college             2   
14         0      76.873174         56    bachelor's degree             0   
15         1      40.515184         26  secondary education             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   

    total_income                              purpose         age_group  
0      40620.102                purchase of the house             adult  
1      17932.802                         car purchase             adult  
2      23341.752                purchase of the house      young adault  
3      42820.568              supplementary education      young adault  
4      25378.572                    to have a wedding  late middle aged  
5      40922.170                purchase of the house      young adault  
6      38484.156                 housing transactions             adult  
7      21731.829                            education  late middle aged  
8      15337.093                     having a wedding             adult  
9      23108.150  purchase of the house for my family             adult  
10     18230.959                      buy real estate             adult  
11     12331.077           buy commercial real estate             adult  
13     20873.317                         car purchase  late middle aged  
14     26420.466          buy residential real estate  late middle aged  
15     18691.345         construction of own property      young adault

In [44]:
# Look at the mean values for income based on your identified factors
print(credit_scoring_full.groupby('education')['total_income'].mean(),
credit_scoring_full.groupby('income_type')['total_income'].mean(),
credit_scoring_full.groupby('family_status')['total_income'].mean(),
credit_scoring_full.groupby('age_group')['total_income'].mean(),
     )

education
bachelor's degree      33197.258790
graduate degree        27960.024667
primary education      21144.882211
secondary education    24594.390815
some college           29028.844227
Name: total_income, dtype: float64 income_type
business                       32434.888945
civil servant                  27336.442546
employee                       25822.872585
paternity / maternity leave     8612.661000
retiree                        21950.722935
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64 family_status
civil partnership    26705.416856
divorced             27245.606570
married              27039.070983
unmarried            26983.707136
widow / widower      22998.392333
Name: total_income, dtype: float64 age_group
adult               28593.480109
late middle aged    25307.312494
senior citizen      21545.431431
teen                16993.942462
young adault        26704.686517
Name: total_income, dtype:

In [45]:
# Look at the median values for income based on your identified factors
print(credit_scoring_full.groupby('education')['total_income'].median(),
credit_scoring_full.groupby('income_type')['total_income'].median(),
credit_scoring_full.groupby('family_status')['total_income'].median(),
credit_scoring_full.groupby('age_group')['total_income'].median(),
     )

education
bachelor's degree      28086.5425
graduate degree        25161.5835
primary education      18741.9760
secondary education    21832.2410
some college           25618.4640
Name: total_income, dtype: float64 income_type
business                       27598.5735
civil servant                  24076.1150
employee                       22815.1035
paternity / maternity leave     8612.6610
retiree                        18959.6260
student                        15712.2600
unemployed                     21014.3605
Name: total_income, dtype: float64 family_status
civil partnership    23195.6360
divorced             23687.2740
married              23369.8045
unmarried            23176.8360
widow / widower      20518.7285
Name: total_income, dtype: float64 age_group
adult               24817.2400
late middle aged    21767.0355
senior citizen      18425.3060
teen                14934.9010
young adault        23432.8590
Name: total_income, dtype: float64


[Repeat such comparisons for multiple factors. Make sure you consider different aspects and explain your thinking process.]

The 4 factors I think have influance on total income are 'education', 'income_type', 'family_status' and 'age_group'. As can be seen from the outputs above 'education' and 'income_type', have far more infuance on the disribution of data. So one of them will be the main factor for the calculation of new values.

[Make a decision on what characteristics define income most and whether you will use a median or a mean. Explain why you made this decision]
The destribution of values by mean() is grater then by median(), and the distrebution by 'income_type' is grater then by 'education'. So I guess filling the missing values by these factors is optimal. 

In [46]:
#  Write a function that we will use for filling in missing values
def fill_income(row):
    income = row['total_income']
    job = row['income_type']
    if  pd.isna(income):
        if job == 'business':
            return 32397.165
        if job == 'civil servant':
            return 27343.730
        if job == 'employee':
            return 25820.841683
        if job == 'paternity / maternity leave':
            return 8612.661000            
        if job == 'retiree':
            return 21940.394503
        if job == 'student':
            return 15712.260000
        if job == 'unemployed':
            return 21014.360500
    else:
        return row['total_income']

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Using absolute values in a function and in general in code is a bad idea. Since the Data Analyst writes code for a specific task, but it can be periodic, which means that the data set will change every time. And in a constant flow of tasks, it will be impossible to track where in the code there is an assignment of absolute values, so always use variables and methods to calculate this or that indicator.
</div>

In [47]:
# Check if it works
credit_scoring_inc = credit_scoring.apply(
    fill_income, axis=1
)
#credit_scoring.head(20)
credit_scoring_inc.head(20)
#credit_scoring.head(20)
#credit_scoring.total_income.isna()

0     40620.102000
1     17932.802000
2     23341.752000
3     42820.568000
4     25378.572000
5     40922.170000
6     38484.156000
7     21731.829000
8     15337.093000
9     23108.150000
10    18230.959000
11    12331.077000
12    21940.394503
13    20873.317000
14    26420.466000
15    18691.345000
16    46272.433000
17    14465.694000
18     9091.804000
19    38852.977000
dtype: float64

In [48]:
# Apply it to every row
credit_scoring['total_income'] = credit_scoring_inc
credit_scoring['total_income'].head(30)

0     40620.102000
1     17932.802000
2     23341.752000
3     42820.568000
4     25378.572000
5     40922.170000
6     38484.156000
7     21731.829000
8     15337.093000
9     23108.150000
10    18230.959000
11    12331.077000
12    21940.394503
13    20873.317000
14    26420.466000
15    18691.345000
16    46272.433000
17    14465.694000
18     9091.804000
19    38852.977000
20    33528.423000
21    21089.953000
22    23948.983000
23    20522.515000
24    46487.558000
25     8818.041000
26    27343.730000
27    49415.837000
28    30058.118000
29    21940.394503
Name: total_income, dtype: float64

In [49]:
# Check if we got any errors
credit_scoring['total_income'].isna().sum()

0

[If you've came across errors in preparing the values for missing data, it probably means there's something special about the data for the category. Give it some thought - you may want to fix some things manually, if there's enough data to find medians/means.]
no erors as far as I can notice.

In [50]:
# Replacing missing values if there are any errors


[When you think you've finished with `total_income`, check that the total number of values in this column matches the number of values in other ones.]

In [51]:
# Checking the number of entries in the columns

credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 13 columns):
children            21230 non-null int64
days_employed       19149 non-null float64
dob_years           21230 non-null int64
education           21230 non-null object
education_id        21230 non-null int64
family_status       21230 non-null object
family_status_id    21230 non-null int64
gender              21230 non-null object
income_type         21230 non-null object
debt                21230 non-null int64
total_income        21230 non-null float64
purpose             21230 non-null object
age_group           21230 non-null object
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Well done! Good job. The gaps in the "total_income" column have been removed.
</div>

###  Restoring values in `days_employed`

[Think about the parameters that may help you restore the missing values in this column. Eventually, you will want to find out whether you should use mean or median values for replacing missing values. You will probably conduct a research similar to the one you've done when restoring data in a previous column.]

In [52]:
# Distribution of `days_employed` medians based on your identified parameters

print(credit_scoring_full.groupby('education')['days_employed'].mean(),
credit_scoring_full.groupby('income_type')['days_employed'].mean(),
credit_scoring_full.groupby('family_status')['days_employed'].mean(),
credit_scoring_full.groupby('age_group')['days_employed'].mean(),
     )


education
bachelor's degree      1768.815729
graduate degree        5055.151259
primary education      5430.851098
secondary education    3187.858027
some college            867.635088
Name: days_employed, dtype: float64 income_type
business                          88.298338
civil servant                    141.338303
employee                          96.905870
paternity / maternity leave      137.364998
retiree                        15209.895255
student                           24.114648
unemployed                     15267.235531
Name: days_employed, dtype: float64 family_status
civil partnership    2450.947731
divorced             2874.352251
married              2646.210865
unmarried            1944.414364
widow / widower      8537.388538
Name: days_employed, dtype: float64 age_group
adult                 409.383421
late middle aged     7018.118378
senior citizen      13082.455952
teen                   26.403254
young adault          103.743743
Name: days_employed, dtype: float

In [53]:
# Distribution of `days_employed` means based on your identified parameters

print(credit_scoring_full.groupby('education')['days_employed'].median(),
credit_scoring_full.groupby('income_type')['days_employed'].median(),
credit_scoring_full.groupby('family_status')['days_employed'].median(),
credit_scoring_full.groupby('age_group')['days_employed'].median(),
     )

education
bachelor's degree       79.023720
graduate degree        235.835710
primary education      126.830567
secondary education     99.749102
some college            50.384599
Name: days_employed, dtype: float64 income_type
business                          64.831141
civil servant                    111.370997
employee                          65.574628
paternity / maternity leave      137.364998
retiree                        15219.545851
student                           24.114648
unemployed                     15267.235531
Name: days_employed, dtype: float64 family_status
civil partnership       81.107055
divorced                99.934334
married                 96.187176
unmarried               60.906619
widow / widower      14037.925329
Name: days_employed, dtype: float64 age_group
adult                  82.755007
late middle aged      384.348805
senior citizen      15020.871042
teen                   30.187192
young adault           49.244340
Name: days_employed, dtype: float

[Decide what you will use: means or medians. Explain why.]
means by age_group groping shows a high spread, but not with giant diferances like the data based on income_type means (which has max values of 15000, in two minor gruopings(retierie and unemployed)) which seams a wee bit unrelible.     

In [54]:
# Let's write a function that calculates means or medians (depending on your decision) based on your identified parameter
credit_scoring_days = credit_scoring.groupby('age_group')['days_employed'].transform(lambda grp: grp.fillna(np.mean(grp)) )
    
    
    
    

In [55]:
# Check that the function works

credit_scoring_days.head(20)

0       351.569709
1       167.700156
2       234.309275
3       171.864467
4     14177.753002
5        38.591076
6       119.966752
7         6.365815
8       288.744387
9        91.198185
10      173.811819
11       33.029245
12    13082.455952
13       76.943414
14       76.873174
15       40.515184
16       71.663926
17       98.749988
18    16678.380705
19      418.284106
Name: days_employed, dtype: float64

In [56]:
# Apply function to the income_type
credit_scoring_days = credit_scoring.groupby('income_type')['days_employed'].transform(lambda grp: grp.fillna(np.mean(grp)) )


In [57]:
# Check if function worked
credit_scoring['days_employed'] = credit_scoring_days
credit_scoring_days.head(20)

0       351.569709
1       167.700156
2       234.309275
3       171.864467
4     14177.753002
5        38.591076
6       119.966752
7         6.365815
8       288.744387
9        91.198185
10      173.811819
11       33.029245
12    15209.895255
13       76.943414
14       76.873174
15       40.515184
16       71.663926
17       98.749988
18    16678.380705
19      418.284106
Name: days_employed, dtype: float64

In [58]:
# Replacing missing values

credit_scoring.head(15) 

children  days_employed  dob_years            education  education_id  \
0          1     351.569709         42    bachelor's degree             0   
1          1     167.700156         36  secondary education             1   
2          0     234.309275         33  secondary education             1   
3          3     171.864467         32  secondary education             1   
4          0   14177.753002         53  secondary education             1   
5          0      38.591076         27    bachelor's degree             0   
6          0     119.966752         43    bachelor's degree             0   
7          0       6.365815         50  secondary education             1   
8          2     288.744387         35    bachelor's degree             0   
9          0      91.198185         41  secondary education             1   
10         2     173.811819         36    bachelor's degree             0   
11         0      33.029245         40  secondary education             1   
12         0   15209.895255         65  secondary education             1   
13         0      76.943414         54         some college             2   
14         0      76.873174         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose         age_group  
0   40620.102000                purchase of the house             adult  
1   17932.802000                         car purchase             adult  
2   23341.752000                purchase of the house      young adault  
3   42820.568000              supplementary education      young adault  
4   25378.572000                    to have a wedding  late middle aged  
5   40922.170000                purchase of the house      young adault  
6   38484.156000                 housing transactions             adult  
7   21731.829000                            education  late middle aged  
8   15337.093000                     having a wedding             adult  
9   23108.150000  purchase of the house for my family             adult  
10  18230.959000                      buy real estate             adult  
11  12331.077000           buy commercial real estate             adult  
12  21940.394503                    to have a wedding    senior citizen  
13  20873.317000                         car purchase  late middle aged  
14  26420.466000          buy residential real estate  late middle aged

[When you think you've finished with `total_income`, check that the total number of values in this column matches the number of values in other ones.]

In [59]:
# Check the entries in all columns - make sure we fixed all missing values
credit_scoring.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_group           0
dtype: int64

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Well done! Good job. The gaps in the "days_employed" column have been removed. And you used a good method to replace it.
</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

~~Please see my comments about the need to correct this column and the proposed method. Please correct this section.~~
</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    I have corrected it, Thanks. 
</div>


## Categorization of data

[To answer the questions and test the hypotheses, you will want to work with categorized data. Look at the questions that were posed to you and that you should answer. Think about which of the data will need to be categorized to answer these questions. Below you will find a template through which you can work your way when categorizing data. The first step-by-step processing covers the text data; the second one addresses the numerical data that needs to be categorized. You can use both or none of the suggested instructions - it's up to you.]

[Despite of how you decide to address the categorization, make sure to provide clear explanation of why you made your decision. Remember: this is your work and you make all decisions in it.]
The 2 categories I think could simolify the data procesing if categorised are:
total_income - by income steps: low income(< 17208.100500), median income(<= 24303.814000), high income(<= 32397.165000), upper class (> 32397.165000)
purpose - real estate, car, wedding, education

In [60]:
credit_scoring['total_income'].describe(include='all').T
#credit_scoring['purpose'].unique()

count     21230.000000
mean      26797.364166
std       15723.514476
min        3306.762000
25%       17208.100500
50%       24303.814000
75%       32397.165000
max      362496.645000
Name: total_income, dtype: float64

In [61]:
# Print the values for your selected data for categorization

credit_scoring['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

[Let's check unique values]

In [62]:
# Check the unique values

[What main groups can you identify based on the unique values?]
real estate, car, wedding, education

[Based on these themes, we will probably want to categorize our data.]


In [63]:
m = Mystem()

english_stemmer = SnowballStemmer('english')
wordnet_lemma = WordNetLemmatizer()

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Please note that it is better to load all libraries at the beginning of any research.
</div>


<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    Thanks, I've moved all of the libraries to the beginning
</div>

In [64]:
lemmas_list_all = []

for purpose in credit_scoring.purpose.unique():
    words = nltk.word_tokenize(purpose)
    lemmas = [wordnet_lemma.lemmatize(w, pos = 'n') for w in words]
    lemmas=[l.lower() for l in lemmas]
    for i in lemmas:
        lemmas_list_all.append(i)

In [65]:
Counter(lemmas_list_all)

Counter({'purchase': 6,
         'of': 5,
         'the': 2,
         'house': 3,
         'car': 9,
         'supplementary': 2,
         'education': 7,
         'to': 6,
         'have': 1,
         'a': 9,
         'wedding': 3,
         'housing': 3,
         'transaction': 4,
         'having': 1,
         'for': 2,
         'my': 4,
         'family': 1,
         'buy': 4,
         'real': 7,
         'estate': 7,
         'commercial': 2,
         'residential': 1,
         'construction': 1,
         'own': 4,
         'property': 4,
         'building': 2,
         'buying': 3,
         'second-hand': 2,
         'with': 2,
         'become': 1,
         'educated': 1,
         'getting': 2,
         'an': 1,
         'ceremony': 1,
         'get': 1,
         'higher': 1,
         'profile': 1,
         'university': 2,
         'renting': 1,
         'out': 1,
         'renovation': 1,
         'going': 1})

In [66]:
real_estate_cat =['house','construction','residential','building','estate','housing','property']
vehicle_cat=['car','cars']
wedding_cat=['ceremony','wedding']
education_cat=['university','education','educated']


In [67]:
# Let's write a function to categorize the data based on common topics
def lemmatization_func(line):
  
    words = nltk.word_tokenize(line)
    lemmas = [wordnet_lemma.lemmatize(w, pos = 'n') for w in words]
    lemmas=[l.lower() for l in lemmas]
    
    if any(word in lemmas for word in real_estate_cat):
        return 'real estate'
    elif  any(word in lemmas for word in vehicle_cat):
        return 'vehicle'
    elif  any(word in lemmas for word in wedding_cat):
        return 'wedding'
    elif  any(word in lemmas for word in education_cat):
        return 'education'
    else:
        return 'other'

In [68]:
sample = credit_scoring.loc[1]['purpose']
lemmatization_func(sample)

'vehicle'

In [69]:
# Create a column with the categories and count the values for them
credit_scoring['purpose_cat']=credit_scoring['purpose'].apply(lemmatization_func)
credit_scoring.head()

children  days_employed  dob_years            education  education_id  \
0         1     351.569709         42    bachelor's degree             0   
1         1     167.700156         36  secondary education             1   
2         0     234.309275         33  secondary education             1   
3         3     171.864467         32  secondary education             1   
4         0   14177.753002         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose         age_group  purpose_cat  
0    purchase of the house             adult  real estate  
1             car purchase             adult      vehicle  
2    purchase of the house      young adault  real estate  
3  supplementary education      young adault    education  
4        to have a wedding  late middle aged      wedding

In [70]:
credit_scoring['purpose_cat'].value_counts()


real estate    10703
vehicle         4258
education       3970
wedding         2299
Name: purpose_cat, dtype: int64

[If you decide to categorize the numerical data, you'll need to come up with the categories for it too.]

In [71]:
# Looking through all the numerical data in your selected column for categorization
credit_scoring['total_income'].unique()

array([40620.102, 17932.802, 23341.752, ..., 14347.61 , 39054.888,
       13127.587])

In [72]:
# Getting summary statistics for the column
credit_scoring['total_income'].describe(include='all').T


count     21230.000000
mean      26797.364166
std       15723.514476
min        3306.762000
25%       17208.100500
50%       24303.814000
75%       32397.165000
max      362496.645000
Name: total_income, dtype: float64

[Decide what ranges you will use for grouping and explain why.]
By income steps: low income(< 17208.100500), median income(<= 24303.814000), high income(<= 32397.165000), upper class (> 32397.165000).
Because it devides the data into 4 neat categories (every 25% of the data is a category) that can give us useful indication of the influance of income on debt repaiment.

In [73]:
# Creating function for categorizing into different numerical groups based on ranges
# I know median income is not the correct term for the category but...
def income_step(income):
    if income < 17208.101:
        return 'low income'
    if income <= 24303.814:
        return 'median income'
    if income <= 32397.165:
        return 'high income'
    return 'upper class'


In [74]:
print(income_step(15000))
print(income_step(20000))
print(income_step(30000))
print(income_step(40000))

low income
median income
high income
upper class


In [75]:
# Creating column with categories
credit_scoring['income_step']= credit_scoring['total_income'].apply(income_step)
credit_scoring.head(15)

children  days_employed  dob_years            education  education_id  \
0          1     351.569709         42    bachelor's degree             0   
1          1     167.700156         36  secondary education             1   
2          0     234.309275         33  secondary education             1   
3          3     171.864467         32  secondary education             1   
4          0   14177.753002         53  secondary education             1   
5          0      38.591076         27    bachelor's degree             0   
6          0     119.966752         43    bachelor's degree             0   
7          0       6.365815         50  secondary education             1   
8          2     288.744387         35    bachelor's degree             0   
9          0      91.198185         41  secondary education             1   
10         2     173.811819         36    bachelor's degree             0   
11         0      33.029245         40  secondary education             1   
12         0   15209.895255         65  secondary education             1   
13         0      76.943414         54         some college             2   
14         0      76.873174         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose         age_group  \
0   40620.102000                purchase of the house             adult   
1   17932.802000                         car purchase             adult   
2   23341.752000                purchase of the house      young adault   
3   42820.568000              supplementary education      young adault   
4   25378.572000                    to have a wedding  late middle aged   
5   40922.170000                purchase of the house      young adault   
6   38484.156000                 housing transactions             adult   
7   21731.829000                            education  late middle aged   
8   15337.093000                     having a wedding             adult   
9   23108.150000  purchase of the house for my family             adult   
10  18230.959000                      buy real estate             adult   
11  12331.077000           buy commercial real estate             adult   
12  21940.394503                    to have a wedding    senior citizen   
13  20873.317000                         car purchase  late middle aged   
14  26420.466000          buy residential real estate  late middle aged   

    purpose_cat    income_step  
0   real estate    upper class  
1       vehicle  median income  
2   real estate  median income  
3     education    upper class  
4       wedding    high income  
5   real estate    upper class  
6   real estate    upper class  
7     education  median income  
8       wedding     low income  
9   real estate  median income  
10  real estate  median income  
11  real estate     low income  
12      wedding  median income  
13      vehicle  me

In [76]:
# Count each categories values to see the distribution
credit_scoring['income_step'].value_counts()

high income      5761
low income       5308
median income    5307
upper class      4854
Name: income_step, dtype: int64

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Well done! The categorization for both columns is correct. Keep it up!
</div>

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [77]:
# Check the children data and paying back on time
credit_scoring.groupby('children')['debt'].mean()

# Calculating default-rate based on the number of children
(credit_scoring.groupby('children')['debt'].mean())*100



children
0    7.545824
1    9.202838
2    9.514468
3    8.231707
4    9.756098
5    0.000000
Name: debt, dtype: float64

**Conclusion**

[Write your conclusions based on your manipulations and observations.]
acording to the data families with 5 children dont tend to get in debt, but sample size of 5 child famelies is relativly small so I think this hypotesys (families with 5 children dont tend to get in debt) needs more reasearch. 
Childless families are also quite reliable (7.546% to default). While families with 2 or 4 chilren are of the reletivly high risk (4 child famelies are the most risky).

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Fine! Good conclusion.
</div>

**Is there a correlation between family status and paying back on time?**

In [78]:
# Check the family status data and paying back on time
credit_scoring.groupby('family_status')['debt'].mean()


# Calculating default-rate based on family status
credit_scoring.groupby('family_status')['debt'].mean()*100


family_status
civil partnership    9.314202
divorced             7.124682
married              7.557521
unmarried            9.784173
widow / widower      6.553911
Name: debt, dtype: float64

**Conclusion**

[Write your conclusions based on your manipulations and observations.]
Clients in civil partnership or unmarried clients, are more likely to default then clients with a different family status. Widows / Widowers are the most reliable. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Right. I agree with you
</div>

**Is there a correlation between income level and paying back on time?**

In [79]:
# Check the income level data and paying back on time
credit_scoring.groupby('income_step')['debt'].mean()


# Calculating default-rate based on income level
credit_scoring.groupby('income_step')['debt'].mean()*100


income_step
high income      8.488110
low income       8.006782
median income    8.818542
upper class      7.045735
Name: debt, dtype: float64

**Conclusion**

[Write your conclusions based on your manipulations and observations.]
upper class clients have the lowest default rate, if the data is analysed by income level. While the lower medium income clients (median income) have the highest default rate. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Logical and correct conclusion!
</div>

**How does credit purpose affect the default rate?**

In [80]:
# Check the percentages for default rate for each credit purpose and analyze them
credit_scoring.groupby('purpose_cat')['debt'].mean()*100


purpose_cat
education      9.294710
real estate    7.259647
vehicle        9.323626
wedding        7.872988
Name: debt, dtype: float64

**Conclusion**

[Write your conclusions based on your manipulations and observations.]
people that get a loan for real estate or a wedding, are less likely to default then those who get a loan for education or to buy a car. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
Good job!
</div>

# General Conclusion 

[List your conclusions in this final section. Make sure you include all your important conclusions you made that led you to the way you processed and analyzed the data. Cover the missing values, duplicates, and possible reasons and solutions for problematic artifacts that you had to address.]
The missing values were symetric between days of employment and income, so I guess that it was due to people who didn't want to disclose that information. The solution for that was using the mean based on place of employment as way to get apropriate values, as the salery is highly dependent on the sphere of employment (buisness, service, manufacture, etc.). The missing values in days of employment could also be replaced that way, but I thought that a mean based on age groups would be more logical as teens and young adults tend to work less then the other age groups. The other problems with those columns are from mistaken entry (hours as days) and prabably system fault (negative work day values).

Duplicates came from human error most likely ("did I enter that data? I don't remmember I'll enter it again"), and could be easily droped but using user ID column is recomanded, that way evevn if 2 clients have similar data they can be identified as diferent entries.

The problematic values in other columns are definatly human error but were insignificant, as lower case and upper  writing could easily be united via proper functions.

[List your conclusions regarding the posed questions here as well.]

My conclusions: 
1. Childless families are quite reliable (7.546% to default). While families with 2 or 4 chilren are of the reletivly high risk (4 child famelies are the most risky).

2. Clients in civil partnership or unmarried clients, are more likely to default then clients with a different family status. Widows / Widowers are the most reliable.

3. Upper class clients have the lowest default rate, if the data is analysed by income level. While the lower medium income clients (median income) have the highest default rate.

4. People that get a loan for real estate purposes or a wedding ceremony, are less likely to default then those who get a loan for education purposes or to buy a car.

<div class="alert alert-block alert-success">
<b>Final reviewer's comment</b> <a class="tocSkip"></a>

You've done a great job! 
    
It was my pleasure to check your work:
1. There is a structure;
2. You add subtotals;
3. Well described the Final conclusions at the end of the project;
4. Correctly removed duplicates and made data categorization;
5. Correctly replaced the missing values in the "total_income" column;
6. He argued his conclusions in the Checking the Hypotheses block.    
</div>

<div class="alert alert-block alert-warning">
<b>Final reviewer's comment</b> <a class="tocSkip"></a>

Please pay attention to a few nuances:
1. It is desirable to add a block describing the Brief Work Plan to the project description;
2. It is better to load all libraries at the very beginning of the project in one cell, and read the project in the next cell;
3. For a quick dive into Python, use the documentation. It will help you optimize your code and take you to the next level.    
</div>

<div class="alert alert-block alert-danger">
<b>Final reviewer's comment</b> <a class="tocSkip"></a>

~~But there are a one remark that need to be completed so that the work can be accepted:~~
~~ - Please fix the method of replacing incorrect values in the "days_employed" column.~~
    
If you have any questions - write, I will definitely help to solve them.
    
I know you will succeed!
</div>

<div class="alert alert-block alert-info">
<b>Final student answer.</b> <a class="tocSkip"></a>
    
I have changed the things that needed to be changed, I realy hope I undersood correctly what was problematic in my work. 
Thanks for the remarks I hope my future work will be better. 
And special thanks for the pandas.DataFrame.query method you recommended, I will surely use it in my future work.      
    
Thanks for all the help, and best wishes.     
</div>

<div class="alert alert-block alert-success">
<b>Final reviewer's comment  2</b> <a class="tocSkip"></a>
Thank you for making changes to the project. You've done excellent research.
</div>